### ISSUE TITLE
Do we have EntityConverter that like ValueConverter?

### ISSUE THREAD

**Issue by codelovercc** [dotnet/efcore#31292]
Do we have `EntityConverter` that like `ValueConverter`? 

Entity class:
```csharp
public class Person
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Secret { get; set; }
    public bool EnCrypted{ get; set; }
}
```
When the `Person.Encrypted` is `true`, we need to encrypt the `Person.Secret`, and the client code does no need to care how to encrypt/decrypt. When write a `Person` instance to database or read it from database, the conditional encryption/decryption will automatic execute.  Like a `ValueConverter` do , and more like a `ConditionalValueConverter` but this class does not exists in EF Core.

I don't know that `ValueConverter` can't do this or not, `ValueConverter` pass the model value to the `convertToProviderExpression`, and the provider value to the `convertFromProviderExpression`, it does not pass the Entity instance. So we don't know the value of `Person.Encrypted` when using `ValueConverter` for conditional encryption/decryption.

1. We can change the class `Person` like this:
```csharp
public class Person
{
    private readonly ICrypter Crypter; 
    public Person(ICrypter crypter)
    {
        Crypter = crypter;
    }
    public int Id { get; set; }
    public string Name { get; set; }
    private string _secret;
    public string Secret 
    {
        get 
        {
            return Encrypted ? Crypter.Decrypt(_secret) : _secret;
        }
        set
        {
            _secret = Encrypted ? Crypter.Encrypt(value) : value;
        }
    }
    private bool _encrypted;
    public bool Encrypted
    {
        get 
        {
            return _encrypted;
        }
        set
        {
            if ( value != _encrypted )
            {
                _secret = value ? Crypter.Encrypt(_secret) : Crypter.Decrypt(_secret);
            }
            _encrypted = value;
        }
    }
}
```
But this will not keep the entity class simple, it need a `ICrypter`.

2. We can add a custom type property on `Person`
```csharp
public class Person
{
    public Person()
    {
    }

    public Person(SecretValue value)
    {
        Value = value;
    }
    public int Id { get; set; }
    public string Name { get; set; }

    // it stored as a text in the database, configured by ModelBuilder.Entity<Person>().Property<SecretValue>("Value").HasConversion(...)
    private SecretValue Value { get; set; }

   // keep this property for query, it also will store in the database.
    public string Secret 
    {
        get 
        {
            return Value.Secret;
        }
        set
        {
            Value.Secret = value;
        }
    }

   // keep this property for query, it also will store in the database.
    public bool Encrypted 
    {
        get 
        {
            return Value.Encrypted;
        }
        set
        {
            Value.Encrypted = value;
        }
    }
}

public class SecretValue
{
    public string Secret { get; set; }
    public bool Encrypted { get; set; }
}
```
Now we can use `ValueConverter` convert the property `Person.SecretValue` value, but we have to save the json of `Person.SecretValue` to the database. I don't want to use JSON column, Sqlite is not supported for now, I don't want save the json text in the database either, I want keep the Table structure simple.

If we have `ConditionalValueConverter` or `EntityConverter` that can access the entity instance in the converter, that would be nicer.

```csharp
public class ConditionalValueConverter<TEntity, TModel, TProvider> 
{
    public ConditionalValueConverter( Expression<Func<TEntity, TModel, TProvider>> convertToProviderExpression,
        Expression<Func<TEntity, TProvider, TModel>> convertFromProviderExpression)
    {
        ...
    }
    ...
}

public class EntityConverter<TEntity, TMappedModel>
{
    public ConditionalValueConverter( Expression<Func<TMappedModel, TEntity>> convertToWrite,
        Expression<Func<TEntity, TMappedModel>> convertFromRead)
    {
        ...
    }
    ...
}
```

`EntityConverter` can be used for map `TEntity` to other type `TMappedModel`, `TMappedModel` can be a `TEntity`.

**Comment by roji** [comment]
@codelovercc there are various ways in which you can already intercept saving and loading of entity instances. For example, you could use [SaveChanges interception](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors#savechanges-interception) to check `Encrypted` before saving, and encrypt your field at the point. On the way back, [materialization interception](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#new-and-improved-interceptors-and-events) can be used to decrypt the value.

**Comment by codelovercc** [comment]
Thank you very mush, I didn't know [materialization interception](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#new-and-improved-interceptors-and-events), I missed this.

**Comment by codelovercc** [comment]
One more question please, Thank you.
If we use [SaveChanges interception](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors#savechanges-interception) to encrypt `Person.Secret`. See the code below:

```csharp
// A method in the class Repository or BuzService
public void AddPerson(Person p)
{
    p.Encrypted = true;
    dbContext.Persons.Add(p);
    dbContext.SaveChanges();
   // when the p is saved, what p.Secret value will be?
    Console.WriteLine(p.Secret);
}
``` 

When the `p` is saved, what `p.Secret` value will be?

**Comment by roji** [comment]
@codelovercc that should be pretty to test.

**Comment by roji** [comment]
You can save the original unencrypted value to some property on the entity and then restore it (the interceptor has SavingChanges for before, SavedChanges for after). An alternative is simply to override SaveChanges() on your DbContext, go through the entities tracked in the change tracker and perform your modifications; then, you can go over them again after calling base.SaveChanges() to restore the values etc.

**Comment by codelovercc** [comment]
After SaveChanges() ,the entity still being tracked right? If I restore the unencrypted value, that property will be marked as modified. I need to undetach it, restore the unencrypted value, detach it. So I prefer the ValueConverter way.

**Comment by roji** [comment]
> So I prefer the ValueConverter way.

No problem, but that only works in the scope of one property. You cannot value-convert based on the value of another property.

**Comment by codelovercc** [comment]
>  I need to undetach it, restore the unencrypted value, detach it.

This may cause unexpected behavior

> 2. We can add a custom type property on Person

I will take this, store a bit of json in the database with column comments it should be understand easier by the others.

Thank you for your answers, now I know we have [materialization interception](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#new-and-improved-interceptors-and-events). 

A suggest: Add the [materialization interception](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#new-and-improved-interceptors-and-events) documents to [interceptors](https://learn.microsoft.com/zh-cn/ef/core/logging-events-diagnostics/interceptors) , I read this document( [interceptors](https://learn.microsoft.com/zh-cn/ef/core/logging-events-diagnostics/interceptors) ) but missed materialization interception.

Wish you have a good day :)

**Comment by roji** [comment]
Yeah, we should indeed document these - opened https://github.com/dotnet/EntityFramework.Docs/issues/4417 (/cc @ajcvickers)
